wpilibj/src/main/java/edu/wpi/first/wpilibj/shuffleboard/Shuffleboard.java

```diff
@@ -0,0 +1,268 @@
+ *   .toTab("Example Tab")
+ *   .withWidget("Toggle Button")
+ *   .getEntry();
+ * }</pre>
+ *
+ * Changing the colors of the boolean box:
+ * <pre>{@code
+ * NetworkTableEntry myBoolean = Shuffleboard.add("My Boolean", NetworkTableType.kBoolean)
+ *   .toTab("Example Tab")
+ *   .withWidget("Boolean Box")
+ *   .withProperties(Map.of("colorWhenTrue", "green", "colorWhenFalse", "maroon"))
+ *   .getEntry();
+ * }</pre>
+ *
+ * Specifying a parent layout. Note that the layout type must <i>always</i> be specified, even if
+ * the layout has already been generated by a previously defined entry.
+ * <pre>{@code
+ * NetworkTableEntry myBoolean = Shuffleboard.add("My Boolean", NetworkTableType.kBoolean)
+ *   .toTab("Example Tab")
+ *   .toLayout("List Layout", "Example List")
+ *   .getEntry();
+ * }</pre>
+ * </p>
+ */
+public final class Shuffleboard {
+
+  public static final String BASE_TABLE_NAME = "/Shuffleboard";
+
+  // Package-private and not final to allow tests to have fresh instances
+  static NetworkTableInstance instance = NetworkTableInstance.getDefault();
+
+  private static List<BuilderBase> m_builders = new ArrayList<>();
+  private static List<SendableData> m_sendables = new ArrayList<>();
+
+  static {
+    HLUsageReporting.reportSmartDashboard(); // TODO use a unique ID for shuffleboard?
+  }
+
+  private static final class SendableData {
+
+    private final SendableBuilder m_sendable;
+    private final SendableBuilderImpl m_builder;
+    private boolean m_initialized = false;
+
+    SendableData(SendableBuilder sendable, SendableBuilderImpl builder) {
+      this.m_sendable = sendable;
+      this.m_builder = builder;
+    }
+
+    /**
+     * Initializes the sendable, if not already initialized. No-op if the sendable has been
+     * initialized.
+     */
+    public void initialize() {
+      if (!m_initialized) {
+        m_builder.setTable(instance.getTable(m_sendable.generateKey()));
+        m_sendable.getSendable().initSendable(m_builder);
+        m_builder.startListeners();
+        m_initialized = true;
+      }
+    }
+
+    /**
+     * Updates the sendable.
+     */
+    public void update() {
+      m_builder.updateTable();
+    }
+  }
+
+  /**
+   * Gets the base Shuffleboard table in the current NetworkTables instance. This is a method
+   * instead of stored as a field because the instance can change during tests, and a field would
+   * still reference the original instance.
+   */
+  private static NetworkTable getBaseTable() {
+    return instance.getTable(BASE_TABLE_NAME);
+  }
+
+  /**
+   * Updates all the values in Shuffleboard. This should be called repeatedly in a loop (for
+   * example, in an iterative or timed robot's autonomousPeriodic and teleopPeriodic methods).
+   */
+  public static void update() {
+    NetworkTable baseTable = getBaseTable();
+    // Update tabs
+    NetworkTable tabsTable = baseTable.getSubTable(".tabs");
+    m_builders.stream()
```

Comment:
Same question about the allocations here, but this one has a ton more. We should try and store these instead of having to reupdate every loop. And should definitely try and avoid some of the spurious allocations if possible.

Reply:
`b -> b.getTab().getName()` is non-capturing, so only one object is allocated.

The lambda in the `forEach` _is_capturing, but could easily be rewritten to be noncapturing. And doing this stream in a conventional loop would be a bit nasty.